msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationKotlinTest.kt

```diff
@@ -2293,3 +2395,33 @@ class NativeAuthPublicClientApplicationKotlinTest : PublicClientApplicationAbstr
         assertTrue((result as SignUpError).isInvalidUsername())
     }
 }
+
+class LoggerTestCallback(private val allowPIITrueToCheck: List<String>, private val allowPIIFalseToCheck: List<String>, private val allowPII: Boolean, private val latch: CountDownLatch?) : ILoggerCallback {
+    var failCalled = false
+    override fun log(
+        tag: String?,
+        logLevel: Logger.LogLevel?,
+        message: String?,
+        containsPII: Boolean
+    ) {
+        if (containsPII) {
+            val elementsToCheck = allowPIITrueToCheck.toMutableList()
+            if (!allowPII) {
+                elementsToCheck.addAll(allowPIIFalseToCheck)
+            }
+            elementsToCheck.forEach { regex ->
+                if (RegexMatcher(regex).matches(message)) {
+                    failCalled = true
+                    fail("PII $regex found in log message: $message")
+                }
+            }
+            latch?.countDown()
+        }
```

Comment:
There's a lot of flags here which make the logic here difficult to follow - I'll add a few assumptions on the code here, correct me if I'm wrong:

* If `containsPII` is false, we should return without checking anything (this could also be debated - should we check the logs even if the function indicates there's no PII? Might help to catch some cases where it's been set incorrectly)
* `allowPIITrueToCheck` is a list of regular expressions that will detect any PII messages that might be present in the logs. If this is the case, what is `allowPIIFalseToCheck` for?

With those assumptions, the following case seems incorrect to me: 
* When `allowPII` is true, the `elementsToCheck` will contain the contents of `allowPIITrueToCheck`. This will mean that if a message contains PII, the regex matcher will return true on a message with PII, and `failCalled` will be set to true. This seems incorrect to me from how the test is laid out. If `allowPII` is true, we shouldn't fail the test, even if the message contain PII.

Could we try and simplify it to only check the messages if `allowPII` is set to false? Is there value in checking log contents when `allowPII` is true?

Reply:
You are right. There is a logical misunderstanding here.

I started out using containPII as a criterion.  If the object doesn't have PII info then the entry is not checked. We only check if we have put invalid attributes for entries that have PII info. However, object.containPII is determined by our log, toString().equals(toUnsanitizedString()). This means that the value of containPII should be used as part of our logging check and not a basis for decision-making.

Reply:
I rename allowPIITrueToCheck as infoPIIToCheck. infoPIIToCheck item should never be logged even when allowPII is true. allowPIIFalseToCheck items are logged or not depending on allowPII.

Reply:
Understood 👍 the updated logic looks good to me, very straightforward to read.

Small suggestion on the naming which might make the intention clearer:
`infoPIIToCheck` -> `sensitivePIIMessages`
`allowPIIFalseToCheck` -> `permittedPIIMessages`

Reply:
Nice suggestions. Updated.